Experimental/src/Ranges/ImmutableRangeList.cs

```diff
@@ -0,0 +1,273 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+using JetBrains.Annotations;
+
+namespace CodeJam.Ranges
+{
+	[PublicAPI]
+	[DebuggerDisplay("{DisplayValue()}")]
+    public class ImmutableRangeList<TValue> : IEnumerable<Range<TValue>>
+        where TValue : IComparable<TValue>
+    {
+        public static readonly ImmutableRangeList<TValue> Empty = new ImmutableRangeList<TValue>();
+        public static readonly ImmutableRangeList<TValue> Full = new ImmutableRangeList<TValue>(Range<TValue>.Full);
+        internal readonly List<Range<TValue>> Values;
+        private int? _hashCode;
+
+        internal ImmutableRangeList(Range<TValue> range)
+        {
+            Values = new List<Range<TValue>> {range};
+        }
+
+        internal ImmutableRangeList()
+        {
+        }
+
+        internal ImmutableRangeList(List<Range<TValue>> values)
+        {
+            Values = values;
+        }
+
+        public bool IsEmpty
+        {
+            get { return Values == null; }
+        }
+
+        public bool IsFull
+        {
+            get
+            {
+                return Values != null
+                       && Values.Count == 1
+                       && Values[0].IsFull;
+            }
+        }
+
+        public IEnumerator<Range<TValue>> GetEnumerator()
+        {
+            if (Values == null)
+                return Enumerable.Empty<Range<TValue>>().GetEnumerator();
+            return Values.GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        protected bool Equals(ImmutableRangeList<TValue> other)
+        {
+            if (IsFull && other.IsFull)
+                return true;
+            if (IsEmpty && other.IsEmpty)
+                return true;
+
+            if (IsEmpty)
+                return false;
+            if (other.IsEmpty)
+                return false;
+
+            if (Values.Count != other.Values.Count)
+                return false;
+
+            for (int i = 0; i < Values.Count; i++)
+            {
+                if (!Equals(Values[i], other.Values[i]))
+                    return false;
+            }
+
+            return true;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj.GetType() != GetType()) return false;
+            return Equals((ImmutableRangeList<TValue>) obj);
+        }
+
+        public override int GetHashCode()
+        {
+            // ReSharper disable NonReadonlyMemberInGetHashCode
+            if (!_hashCode.HasValue)
```

Comment:
I think it will be more heavyweight. It's not really critical here to be thread safe and there is no problem if hash may be calculated twice.